units/src/compact_size.rs

```diff
@@ -0,0 +1,520 @@
+        }
+    }
+
+    /// Returns the number of bytes used to encode this `CompactSize` value.
+    ///
+    /// # Returns
+    ///
+    /// - 1 for 0..=0xFC
+    /// - 3 for 0xFD..=(2^16-1)
+    /// - 5 for 0x10000..=(2^32-1)
+    /// - 9 otherwise.
+    #[inline]
+    pub const fn encoded_size(&self) -> usize {
+        match self.0 {
+            0..=0xFC => 1,
+            0xFD..=0xFFFF => 3,
+            0x10000..=0xFFFFFFFF => 5,
+            _ => 9,
+        }
+    }
+
+    /// Gets the `CompactSize` encoded at the front of `slice`.
+    ///
+    /// Does not modify `slice`, enforces invariant that `size < CompactSize::MAX_SIZE`.
+    ///
+    /// # Returns
+    ///
+    /// The `CompactSize` if one exists as well as the number of bytes used to encode it.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use bitcoin_units::CompactSize;
+    /// let (compact, len) = CompactSize::decode(&[0x20]).expect("single u8 is valid");
+    /// assert_eq!(compact, CompactSize::new(32));
+    /// assert_eq!(len, 1);
+    /// ```
+    pub fn decode(slice: &[u8]) -> Result<(Self, usize), DecodeError> {
+        if slice.is_empty() {
+            return Err(DecodeError::Empty);
+        }
+
+        let res = match slice[0] {
+            0xFF => {
+                const SIZE: usize = 9;
+                let len = slice.len();
+                if len < SIZE {
+                    return Err(DecodeError::TooShort(TooShortError::new(slice, SIZE)));
+                }
+                let mut bytes = [0_u8; SIZE - 1];
+                bytes.copy_from_slice(&slice[1..SIZE]);
+                let v = u64::from_le_bytes(bytes);
+                Ok((CompactSize::new(v.into()), SIZE))
+            },
+            0xFE => {
+                const SIZE: usize = 5;
+                let len = slice.len();
+                if len < SIZE {
+                    return Err(DecodeError::TooShort(TooShortError::new(slice, SIZE)));
+                }
+                let mut bytes = [0_u8; SIZE - 1];
+                bytes.copy_from_slice(&slice[1..SIZE]);
+                let v = u32::from_le_bytes(bytes);
+                Ok((CompactSize::new(v.into()), SIZE))
+            },
+            0xFD => {
+                const SIZE: usize = 3;
+                let len = slice.len();
+                if len < SIZE {
+                    return Err(DecodeError::TooShort(TooShortError::new(slice, SIZE)));
+                }
+                let mut bytes = [0_u8; SIZE - 1];
+                bytes.copy_from_slice(&slice[1..SIZE]);
+                let v = u16::from_le_bytes(bytes);
+                Ok((CompactSize::new(v.into()), SIZE))
+            },
+            n => Ok((CompactSize(n.into()), 1))
+        };
+        res
+    }
+
+    /// Encodes `CompactSize` as a vector of bytes.
+    ///
+    /// The returned vector is the exact size of the encoded value.
+    #[cfg(feature = "alloc")]
+    pub fn encode(&self) -> alloc::vec::Vec<u8> {
```

Comment:
Just make this `ArrayVec` in `internals`.

Reply:
I removed this function and returned a `[8u; 9]`.